isobus/src/isobus_virtual_terminal_client_state_tracker.cpp

```diff
@@ -0,0 +1,107 @@
+
+#include <algorithm>
+
+namespace isobus
+{
+	VirtualTerminalClientStateTracker::VirtualTerminalClientStateTracker(std::shared_ptr<ControlFunction> client) :
+	  client(client)
+	{
+	}
+
+	void VirtualTerminalClientStateTracker::initialize()
+	{
+		CANNetworkManager::CANNetwork.add_any_control_function_parameter_group_number_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::VirtualTerminalToECU), process_rx_message, this);
+	}
+
+	void VirtualTerminalClientStateTracker::add_tracked_numeric_value(std::uint16_t objectId, std::uint32_t initialValue)
+	{
+		if (numericValueStates.find(objectId) != numericValueStates.end())
+		{
+			CANStackLogger::warn("[VTStateHelper] add_tracked_numeric_value: objectId %lu already tracked", objectId);
+			return;
+		}
+
+		numericValueStates[objectId] = initialValue;
+	}
+
+	void VirtualTerminalClientStateTracker::remove_tracked_numeric_value(std::uint16_t objectId)
+	{
+		if (numericValueStates.find(objectId) == numericValueStates.end())
+		{
+			CANStackLogger::warn("[VTStateHelper] remove_tracked_numeric_value: objectId %lu was not tracked", objectId);
+			return;
+		}
+
+		numericValueStates.erase(objectId);
+	}
+
+	std::uint32_t VirtualTerminalClientStateTracker::get_numeric_value(std::uint16_t objectId) const
+	{
+		if (numericValueStates.find(objectId) == numericValueStates.end())
+		{
+			CANStackLogger::warn("[VTStateHelper] get_numeric_value: objectId %lu not tracked", objectId);
+			return 0;
+		}
+
+		return numericValueStates.at(objectId);
+	}
+
+	void VirtualTerminalClientStateTracker::process_rx_message(const CANMessage &message, void *parentPointer)
+	{
+		auto *parent = static_cast<VirtualTerminalClientStateTracker *>(parentPointer);
+		parent->process_rx_message(message);
+	}
+
+	void VirtualTerminalClientStateTracker::process_rx_message(const CANMessage &message)
+	{
+		if (message.has_valid_source_control_function() &&
+		    message.is_destination(client) &&
+		    message.is_parameter_group_number(CANLibParameterGroupNumber::VirtualTerminalToECU))
+		{
+			std::uint8_t function = message.get_uint8_at(0);
+			switch (function)
+			{
+				case static_cast<std::uint8_t>(VirtualTerminalClient::Function::ChangeNumericValueCommand):
+				{
+					auto errorCode = message.get_uint8_at(3);
+					if (errorCode == 0)
+					{
+						std::uint16_t objectId = message.get_uint16_at(1);
+						if (numericValueStates.find(objectId) != numericValueStates.end())
+						{
+							std::uint32_t value = message.get_uint32_at(4);
+							numericValueStates[objectId] = value;
+						}
+					}
+					break;
+				}
+				case static_cast<std::uint8_t>(VirtualTerminalClient::Function::VTChangeNumericValueMessage):
+				{
+					std::uint16_t objectId = message.get_uint16_at(1);
+					if (numericValueStates.find(objectId) != numericValueStates.end())
+					{
+						std::uint32_t value = message.get_uint32_at(4);
+						numericValueStates[objectId] = value;
+					}
+					break;
+				}
```

Comment:
The VT's not obligated to change the displayed value when it sends the ChangeNumericValueCommand message, since it relies on the client somewhat to ensure that the value which was set is valid, so when this happens, if we really want to be helpful, we may want to also transmit our version of the command back to the VT

Reply:
Hmm I agree, though we currently already send a response back in the virtual terminal client class: https://github.com/Open-Agriculture/AgIsoStack-plus-plus//blob/daan/vt-track-and-update-states/isobus/src/isobus_virtual_terminal_client.cpp#L2559-L2578.

Do you think we should change that? I tried to made the tracker class standalone such that it in theory can track any VT client, not just ours. Hence sending responses in here is a little off I think

Reply:
Also the response value will be identical to the value per definition of the standard, and if there's no response from the client the VT will (for version 5 and later) act as if an unexpected shutdown occurred. Maybe the latter is not really incorporated yet into the state tracker, but it's a bit out-of-scope for this PR I feel like. I'll add a TODO for it :)

Reply:
Okay I delved a bit deeper into this, we might instead need to add a way of entering a so called "safe state" in the `VirtualTerminalClient` itself. When an unexpected shutdown occurs, the standard requires us to put all functions dependant on the VT operator interface into a known state such that it will not put the operator or machine at risk. I think we should definitely add proper implementation for that instead of just restarting the connection as we do now

Reply:
> though we currently already send a response back

Right, but that's not the same thing as commanding the value. I think sending the response there is appropriate.

> I tried to made the tracker class standalone such that it in theory can track any VT client, not just ours. Hence sending responses in here is a little off I think

To clarify - I'm not talking about the response, I mean we should send the non-VT version of the change numeric value command in the helper. This is important with older VTs, like version 4, which may not care about the responses.

> we might instead need to add a way of entering a so called "safe state"

Only if entering a safe state is a callback to the user's app in some way. If a shutdown occurs, the VT client state we have is meaningless - rather, it's up to the main app (the person integrating our library) to know what their implement is doing and take some action accordingly. This is not something we can control outside of some kind of notification to the user.

Reply:
> I mean we should send the non-VT version of the change numeric value command in the helper

Oh hmm yeah, you mean we should send a reply to the VT with a command message (0xA8)? Is it concerning the race condition that can occur where both the VT and working set apply a change at the same time? Would that be something we can also do in the VT client interface, I'm still unsure if the tracker is the right place for that

> it's up to the main app (the person integrating our library) to know what their implement is doing and take some action accordingly. This is not something we can control outside of some kind of notification to the user.

For sure, I indeed meant it in a way where the application indeed get's notified and can take action accordingly. Restarting the connection blindly like we do now can introduce some concerning side effects I imagine

Reply:
> you mean we should send a reply to the VT with a command message (0xA8)

Yeah, that one! We could maybe make these proposed A8 commands optional? But it seems like it could be a nice quality of life feature in the general sense.

> Would that be something we can also do in the VT client interface, I'm still unsure if the tracker is the right place for that

I don't think the client is the ideal place, because it's up to the application/the user to take the value set by the VT and figure out if that value is "valid" or even if they want to do some other behavior. That's kind of why I was think the helper made more sense, it's like a way to make the default behavior to accept whatever the VT says is happening. What do you think?

> Is it concerning the race condition that can occur where both the VT and working set apply a change at the same time?

Somewhat? We should think about that a bit probably, though I am not sure how I would prevent it... 

> For sure, I indeed meant it in a way where the application indeed get's notified and can take action accordingly. 

OK perfect, I can add some thoughts to that issue as well

Reply:
> it's up to the application/the user to take the value set by the VT and figure out if that value is "valid" or even if they want to do some other behavior. That's kind of why I was think the helper made more sense, it's like a way to make the default behavior to accept whatever the VT says is happening. What do you think?

Ahh okay now I understand. I agree that the client interface is not the best place providing a complete interface for that. Though the implementation you are trying to achieve is almost there already; We have the `add_vt_change_numeric_value_event_listener` in the client interface, and using that in combination with the `send_change_numeric_value` should yield your result. For simplicity this is already quite good, what do you think?

Let's take our seeder app as an example - I'm in the process of converting it to the helper as well:
```c++
void SeederVtApplication::handle_numeric_value_events(const isobus::VirtualTerminalClient::VTChangeNumericValueEvent &event)
{
	switch (event.objectID)
	{
		case statisticsSelection_VarNum:
		{
			auto newSelectedStatistic = static_cast<Statistics>(event.value);
			if (newSelectedStatistic != currentlySelectedStatistic)
			{
				currentlySelectedStatistic = newSelectedStatistic;
				txFlags.set_flag(static_cast<std::uint32_t>(UpdateVTStateFlags::UpdateStatisticsSelection_VarNum));
				txFlags.set_flag(static_cast<std::uint32_t>(UpdateVTStateFlags::UpdateSelectedStatisticsContainer_ObjPtr));
			}
		}
		break;

		case enableAlarms_VarNum:
		{
			alarmsEnabled = event.value;
			txFlags.set_flag(static_cast<std::uint32_t>(UpdateVTStateFlags::UpdateEnableAlarms_VarNum));
		}
		break;

		default:
			break;
	}
}

``` 

can be reduced to - while providing the same functionality:

```c++
void SeederVtApplication::handle_numeric_value_events(const isobus::VirtualTerminalClient::VTChangeNumericValueEvent &event)
{
	switch (event.objectID)
	{
		case statisticsSelection_VarNum:
		{
			// Acknowledge the selection change by sending it back to the VT
			VTClientInterface->send_change_numeric_value(statisticsSelection_VarNum, event.value);

			// Update the frame to show the new selected statistic
			std::uint16_t newContainer = isobus::VirtualTerminalClient::NULL_OBJECT_ID;
			auto correspondingContainer = STATISTICS_CONTAINER_MAP.find(static_cast<Statistics>(event.value));
			if (STATISTICS_CONTAINER_MAP.end() != correspondingContainer)
			{
				newContainer = correspondingContainer->second;
			}
			VTClientUpdateHelper.set_numeric_value(selectedStatisticsContainer_ObjPtr, newContainer);
		}
		break;

		case enableAlarms_VarNum:
		{
			// Acknowledge the selection change by sending it back to the VT
			VTClientInterface->send_change_numeric_value(enableAlarms_VarNum, event.value);
		}
		break;

		default:
			break;
	}
}
```

Reply:
Or do you reckon registering an extra callback inside the tracker class is more clear? I'm not sure what the default behavior should then be if one would not provide the callback. hmm...

Reply:
Yeah I guess I'm not really sure... the example you posted above is probably simple enough? But like, if we did register an extra callback in the tracker, I'd expect the entirety of `case enableAlarms_VarNum:` could even be done away with. IDK if that's better, or weirder haha

Reply:
I agree. It is probably better to provide it inside the tracker indeed, I can add it. Should be pretty easy now I think about it hahah